### PR TITLE
Add ability to spawn pooled queues with dedicated worker procs.

### PR DESCRIPTION
**Pull request by JeffLuckett**
Current functionality allows for spawning a number of workers to service an array of Queues and have them run under a single process.

I had the need to be able to spawn multiple queues under the Master process with dedicated numbers of workers per queue (as opposed to the default x workers servicing y queues).

If you're interested in adding this functionality to your gem, I'd love your feedback on my implementation, and what you feel I need to do to make this PR "merge worthy".

### CODE REVIEW

**Comment by jturkel** [comment]
@JeffLuckett - I would definitely be open to adding this functionality. I think the biggest thing to work out is how this fits in with the configuration DSL. Ideally it would support configuring the sleep delay and number of worker processes per "worker type" while still making the common case of a single worker type easy.

**Comment by JeffLuckett** [comment]
@jturkel - Sorry for missing your comment!  I didn't get an email notification for some reason.

The implementation I provided simply passes along all valid parameters to the DJ workers.  So you can easily configure any "valid" parameter that way.

Also, my functionality is additive.  You can still invoke the master process with your original config, and it works just like it always did ... and can even do so alongside my pooled queues functionality.

While we haven't done so yet, we're planning to utilize that functionality to run a cron-like queue for stuff we need to have done periodically and automatically without having to actually use cron on the server.

Here's a gist containing my config files: https://gist.github.com/JeffLuckett/609bdc717960cb04254f97ba48ec3b50

Basically, anything you put into `dj_opts` in that YAML file gets passed to the workers as they're spawned.  It gives you some pretty fine-grained control.

**Comment by JeffLuckett** [comment]
I didn't include this in my PR, but in my own project, I created some rake tasks that can be used either in development or as part of a deploy, etc...

https://gist.github.com/JeffLuckett/86407fdc5bf36c6d5ca19d304b851310

This creates tasks that start individual queues "inline" and are useful for development/debugging etc...  These are dynamically generated from the .yaml file in the other gist.

They create rake tasks that are invoked like `bundle exec rake jobs:start:[queue_name]`

Then there's the `jobs:start:all` task that we use to invoke the `DelayedJobWorkerPool`.  This is used in Production where we're running the worker proc in a Docker container.

**Comment by letrungkien** [comment]
@JeffLuckett Hey man, thanks for the awesome feature, I'm in for this implementation. Can you update the README as well please? :bowtie:

**Comment by severinraez** [comment]
Now I'm curious...  https://github.com/voyagin/delayed_job_worker_pool/pull/1 looks like a cleaned up version of this. Is the gem still used/maintained, and if yes can you merge either one, @jturkel ?

**Comment by jturkel** [comment]
This gem is still maintained. It's been stable for a long time so fortunately it hasn't required much work :) 

I'm definitely open to adding this functionality to the gem but I think the proposed configuration DSL needs a bit of work. Namely:
* We should use configuration methods (e.g. `min_priority`) vs. hashes to configure the additional worker pools for consistency with the existing DSL.
* The relationship between the main pool and the additional worker pools isn't very obvious - are they just separate pools? are they separate pools + defaults? Better naming could help here but the cleanest way to expose this might be a separate Rake task + configuration DSL where there is no default pool e.g.
```ruby
worker_pools do
  worker_pool do
    workers 5
    queues: 'my-queues'
  end

  worker_pool do
    workers 10
    queues: 'my-other-queues'
  end
end
```
I'd be happy to review a PR if someone is interested in pushing this forward.

**Comment by severinraez** [comment]
That was my strong suspection regarding the gem health ;)
Thanks for clarifying your ideas, sounds reasonable. I might find time to give this a go in the next months (if nobody beats me to it :P)

**Comment by severinraez** [comment]
Hi @jturkel

I've drafted an implementation of the DSL you suggested: https://github.com/severinraez/delayed_job_worker_pool/commit/d02d9fc314e3b6a6d412b644403fdf6dfdd812a1
 I am not too happy with it. Before I try to find a cleaner approach, I'd like to get your thoughts on some points.

Goals:

* I tried to accept either "implicit" or "explicit" worker pool definitions (i.e. old or new DSL).
* I wanted to avoid nested instance_eval calls so I don't have to take care of keeping variables from outside the evaled block in scope (by default you loose access to them because you're in the class instance binding during eval)

Problems:

* It's fragile (correct nesting of the dsl is checked rather implicitly)
* Internal methods are leaking out into the DSL
* Concerns are mixed (validation vs DSL)

But, taking a step back: Would you be opposed to making a breaking change and allowing `worker_pools` style configuration only? The upgrade path would be easy, users would have to surround their existing configurations with `worker_pool`. Upside: Gem scope stays small, there's only one way to do configuration.

Regardless of the way you want to move forward here, I propose this syntax to avoid nested `instance_eval` (surrounded by `worker_pools { |pools| ... }` if we stil accept implicit worker pool definitions):

```ruby
worker_pool do |p|
  p.workers 5
  p.queues: 'my-queues'
end

worker_pool do |p|
  p.workers 10
  p.queues: 'my-other-queues'
end
```

**Comment by jturkel** [comment]
@severinraez - Sorry for the delay. Life has been a little crazy lately. I'm not opposed to making a breaking change to the DSL as long as the single worker pool case (which I suspect is the most common one) is still easy enough. How would the callbacks fit into the `worker_pool` configuration blocks? Would they still be done at the top level?

**Comment by severinraez** [comment]
@jturkel - No worries. Took me 7 months to get back to this in the first place ;)

Yes, the idea was to keep callback configurations toplevel. I can't see a reason why they would be pool specific.

I'll try to come up with something then, using the proposed new DSL.

**Comment by jturkel** [comment]
@severinraez - Sounds good to me. I suspect you're right that most callback configuration will be done at the top level. We can always extend the model later to support per-pool callbacks if necessary.

**Comment by jturkel** [comment]
This functionality has been added in #7.
